### PR TITLE
Refactor of producer module (preview)

### PR DESCRIPTION
**Pull request by jaredjj3**
Hey @rvilarl, I've been cleaning up the producer modules and I developed this new pattern that prioritizes testability and encapsulation. I haven't wired it into the main flow of things, so anything you're working on should be unaffected for now.

We can now reap the benefits that `VexmlMessage` (formerly `EasyScoreMessage`) gives us:

- We can develop producer and rendering features separately.
- It's easier to test the producers using POJOs instead of evaluating image diffs or evaluating the equality of complex objects.

Let me know what you think. If you like it, I'll set up Github actions so we can move towards publishing this library.

### CODE REVIEW

File: src/attributes.ts [hunk 1]
```diff
@@ -14,7 +14,7 @@ export class Attributes {
     cur1stStave: number,
     duration: number,
     notes: Array<VF.Note>,
-    systems: VF.System[]
+    systems: System[]
```
Comment (rvilarl) [inline src/attributes.ts:17]: Why is System no longer VF.System and Note stays as VF.Note?

File: src/notations.ts [hunk 2]
```diff
@@ -99,7 +99,7 @@ export class Notations {
           }
           if (modifier.class == 'O') {
             notes[notes.length - 1].addModifier(
-              factory.Ornament(modifier.type, { position: message.placement ?? 'above' }),
+              (factory as any).Ornament(modifier.type, { position: message.placement ?? 'above' }),
```
Comment (rvilarl) [inline src/notations.ts:102]: Why is 'as any' now required?

File: src/types.ts [hunk 3]
```diff
@@ -107,3 +117,14 @@ export interface CodeTracker {
...
+// TODO(jared): Remove hack by contributing to the types upstream.
+export type System = VF.System & {
+  getStaves(): VF.Stave[];
+  getX(): number;
+  setX(x: number): void;
+  getY(): number;
+  setY(y: number): void;
+  addVoices(voices: VF.Voice[]): void;
+  appendSystem(width: number): System;
+};
```
Comment (rvilarl) [inline src/types.ts:130]: Why is that necessary?

**Review by rvilarl — COMMENTED** [review]
Hi Jared,
I trust that the change is good, I do not get it 100% but go for it! I comment on a few things as well.
Cheers, Rodrigo

File: src/types.ts (hunk 3 above)
Comment (jaredjj3) [reply src/types.ts:130]: `VF.System` needs to be updated to include the additional types. Without this, this causing type errors.

File: src/notations.ts (hunk 2 above)
Comment (jaredjj3) [reply src/notations.ts:102]: To workaround type errors.

File: src/attributes.ts (hunk 1 above)
Comment (jaredjj3) [reply src/attributes.ts:17]: VF.System needs to be updated to include the additional types. Without this, there will be type errors.

File: src/types.ts (hunk 3 above)
Comment (rvilarl) [reply src/types.ts:130]: All you need to do is to install the vexflow pre release under ./vexflow/

Comment (rvilarl) [reply src/types.ts:130]: `yarn add ./vexflow/vexflow-5fa72362.tgz`

Comment (rvilarl) [reply src/types.ts:130]: This might also resolve the other two comments

**Comment by jaredjj3** [comment]
> Hi Jared, I trust that the change is good, I do not get it 100% but go for it! I comment on a few things as well. Cheers, Rodrigo

In the current `Producer`, you need to coordinate state inside the class and extend `Producer.getMessages` to handle new nodes. This is hard to maintain, especially at the rate we're going. I made the following changes:

- Move general state into a `ctx` variable that can be passed around. I think this will be easier to reason about and test, since you don't have to prime the class with certain state.
- Create the concept of node handlers that translate `Node` objects into `VexmlMessages`. This makes the inputs/outputs clear: given a node, you expect certain messages to be produced. This makes it easy to test. At first, I wanted to do simple functions, but it got messy trying to fit all the logic in a simple function.
- Use a lightweight dependency injection system by letting the child handlers be passed into the constructor. This makes it possible to truly isolate testing a given handler — instead of injecting a real handler, you inject a `NoopHandler` and then use the testing framework to spy/mock out the methods of interest.

The theme I was going for was encapsulation and testability. Hopefully, it'll make more sense as I add more handlers.
